internal/module/kernelmapper.go

```diff
@@ -104,41 +104,63 @@ func (k *kernelMapper) findKernelMapping(mappings []kmmv1beta1.KernelMapping, ke
 	return nil, errors.New("no suitable mapping found")
 }
 
-func (k *kernelMapper) GetNodeOSConfig(node *v1.Node) *NodeOSConfig {
-	return k.GetNodeOSConfigFromKernelVersion(node.Status.NodeInfo.KernelVersion)
-}
+func (kh *kernelMapperHelper) prepareMappingData(mapping *kmmv1beta1.KernelMapping, modSpec *kmmv1beta1.ModuleSpec, kernelVersion string) error {
+	var err error
 
-func (k *kernelMapper) GetNodeOSConfigFromKernelVersion(kernelVersion string) *NodeOSConfig {
-	osConfig := NodeOSConfig{}
+	// prepare the build
+	if mapping.Build != nil || modSpec.ModuleLoader.Container.Build != nil {
+		mapping.Build = kh.buildHelper.GetRelevantBuild(modSpec.ModuleLoader.Container.Build, mapping.Build)
+	}
 
-	osConfigFieldsList := regexp.MustCompile("[.,-]").Split(kernelVersion, -1)
+	// prepare the sign
+	if mapping.Sign != nil || modSpec.ModuleLoader.Container.Sign != nil {
+		mapping.Sign, err = kh.signHelper.GetRelevantSign(modSpec.ModuleLoader.Container.Sign, mapping.Sign, kernelVersion)
+		if err != nil {
+			return fmt.Errorf("failed to get the relevant Sign configuration for kernel %s: %v", kernelVersion, err)
+		}
+	}
+
+	// prepare TLS options
+	if mapping.RegistryTLS == nil {
+		mapping.RegistryTLS = &modSpec.ModuleLoader.Container.RegistryTLS
+	}
 
-	osConfig.KernelFullVersion = kernelVersion
-	osConfig.KernelVersionMMP = strings.Join(osConfigFieldsList[:kernelVersionPatchIdx+1], ".")
-	osConfig.KernelVersionMajor = osConfigFieldsList[kernelVersionMajorIdx]
-	osConfig.KernelVersionMinor = osConfigFieldsList[kernelVersionMinorIdx]
-	osConfig.KernelVersionPatch = osConfigFieldsList[kernelVersionPatchIdx]
+	//prepare container image
+	if mapping.ContainerImage == "" {
+		mapping.ContainerImage = modSpec.ModuleLoader.Container.ContainerImage
+	}
 
-	return &osConfig
+	return nil
 }
 
-func (k *kernelMapper) PrepareKernelMapping(mapping *kmmv1beta1.KernelMapping, osConfig *NodeOSConfig) (*kmmv1beta1.KernelMapping, error) {
-	osConfigStrings := k.prepareOSConfigList(*osConfig)
+func (kh *kernelMapperHelper) replaceTemplates(mapping *kmmv1beta1.KernelMapping, osConfig *nodeOSConfig) error {
+	var err error
+	osConfigStrings := prepareOSConfigList(*osConfig)
 
 	parser := parse.New("mapping", osConfigStrings, &parse.Restrictions{})
 
-	substContainerImage, err := parser.Parse(mapping.ContainerImage)
+	mapping.ContainerImage, err = parser.Parse(mapping.ContainerImage)
 	if err != nil {
-		return nil, fmt.Errorf("failed to substitute the os config into ContainerImage field: %w", err)
+		return fmt.Errorf("failed to substitute the os config into ContainerImage field: %w", err)
 	}
+	return nil
+}
+
+func (kh *kernelMapperHelper) getNodeOSConfigFromKernelVersion(kernelVersion string) *nodeOSConfig {
+	osConfig := nodeOSConfig{}
 
-	substMapping := mapping.DeepCopy()
-	substMapping.ContainerImage = substContainerImage
+	osConfigFieldsList := regexp.MustCompile("[.,-]").Split(kernelVersion, -1)
```

Comment:
`regexp.MustCompile` can panic, we should extract it from the function and make it a package variable.